src/pages/workspaces/workspace/delete-workspace-modal.state.test.ts

```diff
@@ -0,0 +1,250 @@
+import { act, renderHook } from '@testing-library/react-hooks'
+import { Ajax } from 'src/libs/ajax'
+import { reportError } from 'src/libs/error'
+import { DeepPartial } from 'src/libs/type-utils/deep-partial'
+import { AzureWorkspaceInfo, GoogleWorkspaceInfo } from 'src/libs/workspace-utils'
+import { useDeleteWorkspaceState } from 'src/pages/workspaces/workspace/delete-workspace-modal.state'
+import { asMockedFn } from 'src/testing/test-utils'
+
+
+type AjaxExports = typeof import('src/libs/ajax')
+jest.mock('src/libs/ajax', (): AjaxExports => {
+  return {
+    ...jest.requireActual('src/libs/ajax'),
+    Ajax: jest.fn()
+  }
+})
+
+jest.mock('src/libs/error', () => ({
+  ...jest.requireActual('src/libs/error'),
+  reportError: jest.fn(),
+}))
+
+type AjaxContract = ReturnType<typeof Ajax>;
+type AjaxAppsContract = AjaxContract['Apps']
+type AjaxWorkspaceManagerContract = AjaxContract['WorkspaceManagerResources']
+type AjaxWorkspacesContract = AjaxContract['Workspaces']
+
+describe('useDeleteWorkspace', () => {
+  const googleWorkspace = {
+    accessLevel: 'writer',
+    canShare: true,
+    canCompute: true,
+    workspace: {
+      name: 'example',
+      namespace: 'example',
+      cloudPlatform: 'Gcp'
+    } as GoogleWorkspaceInfo
+  }
+  const azureWorkspace = {
+    accessLevel: 'writer',
+    canShare: true,
+    canCompute: true,
+    workspace: {
+      name: 'example',
+      namespace: 'example',
+      cloudPlatform: 'azure'
+    } as AzureWorkspaceInfo
+  }
+  const mockOnDismiss = jest.fn(() => {
+  })
+  const mockOnSuccess = jest.fn(() => {
+  })
+
+  beforeEach(() => {
+    jest.spyOn(console, 'error').mockImplementation(() => {
+    })
+    jest.spyOn(console, 'log').mockImplementation(() => {
+    })
+  })
+
+  it('can initialize state for a google workspace with running apps', async () => {
+    // Arrange
+    const mockApps: Partial<AjaxAppsContract> = {
+      listWithoutProject: jest.fn()
+    }
+    asMockedFn((mockApps as AjaxAppsContract).listWithoutProject).mockResolvedValue([{
+      appName: 'app1',
+      status: 'running'
+    }])
+
+    const mockGetAcl = jest.fn().mockResolvedValue({ acl: { 'example1@example.com': {} } })
+    const mockGetBucketUsage = jest.fn().mockResolvedValue({ usageInBytes: 1234 })
+    const mockWorkspaces: DeepPartial<AjaxWorkspacesContract> = {
+      workspace: () => ({
+        getAcl: mockGetAcl,
+        bucketUsage: mockGetBucketUsage
+      })
+    }
+
+    const mockAjax: Partial<AjaxContract> = {
```

Comment:
this arrange section, and others like it, could probably be streamlined now that we have DeepPartial type available.  Try replacing line 63 onwards with:

```
const mockAjax:  DeepPartial<AjaxContract> = {
  Apps: {
    listWithoutProject: jest.fn()
  },
  Workspaces: {
    workspace: {
      getAc1: jest.fn()
      bucketUsage: jest.fn()
    }
  }
}
asMockedFn(mockAjax.Apps.listWithoutProject).mockResolvedValue([.....])
asMockedFn(mockAjax.Workspaces.workspace.getAc1).mockResolvedValue({.....})
asMockedFn(mockAjax.Workspaces.workspace.bucketUsage).mockResolvedValue({.....})

asMockedFn(Ajax).mockImplementation(() => mockAjax as AjaxContract)

```
...not dramatically shorter, but at least a bit less jumbled

Reply:
oh wait... never mind... workspace is an arrow function that returns it's own sub-contract object.... uggg.

Reply:
ok, yeah... scratch all that and maybe consider the WorkspaceProvider pattern we discussed and how it would greatly simplify your mocking surface in the unit tests here.